bridges/snowbridge/pallets/system-frontend/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+pub enum ControlCall {
+	#[codec(index = 1)]
+	CreateAgent { location: Box<VersionedLocation>, fee: u128 },
+	#[codec(index = 2)]
+	RegisterToken { asset_id: Box<VersionedLocation>, metadata: AssetMetadata, fee: u128 },
+}
+
+#[allow(clippy::large_enum_variant)]
+#[derive(Encode, Decode, Debug, PartialEq, Clone, TypeInfo)]
+pub enum SnowbridgeControl {
+	#[codec(index = 90)]
+	Control(ControlCall),
+}
+
+#[cfg(feature = "runtime-benchmarks")]
+pub trait BenchmarkHelper<O>
+where
+	O: OriginTrait,
+{
+	fn make_xcm_origin(location: Location) -> O;
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Origin check for XCM locations that can create agents
+		type CreateAgentOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Location>;
+
+		/// Origin check for XCM locations that can create agents
+		type RegisterTokenOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Location>;
+
+		/// XCM message sender
+		type XcmSender: SendXcm;
+
+		/// To withdraw and deposit an asset.
+		type AssetTransactor: TransactAsset;
+
+		#[cfg(feature = "runtime-benchmarks")]
+		type Helper: BenchmarkHelper<Self::RuntimeOrigin>;
+
+		type FeeAsset: Get<Location>;
+
+		type DeliveryFee: Get<Asset>;
+
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A `CreateAgent` message was sent to Bridge Hub
+		CreateAgent { location: Location },
+		/// A message to register a Polkadot-native token was sent to Bridge Hub
+		RegisterToken {
+			/// Location of Polkadot-native token
+			location: Location,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Convert versioned location failure
+		UnsupportedLocationVersion,
+		/// Check location failure, should start from the dispatch origin as owner
+		OwnerCheck,
+		/// Send xcm message failure
+		Send,
+		/// Withdraw fee asset failure
+		FundsUnavailable,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Call `create_agent` to instantiate a new agent contract representing `origin`.
+		/// - `origin`: Can be any sovereign `Location`
+		/// - `fee`: Fee in Ether paying for the execution cost on Ethreum
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::create_agent())]
+		pub fn create_agent(origin: OriginFor<T>, fee: u128) -> DispatchResult {
+			let origin_location = T::CreateAgentOrigin::ensure_origin(origin)?;
+
+			Self::burn_fees(origin_location.clone(), fee)?;
+
+			let call = SnowbridgeControl::Control(ControlCall::CreateAgent {
+				location: Box::new(VersionedLocation::from(origin_location.clone())),
+				fee,
+			});
+
+			let xcm: Xcm<()> = vec![
+				ReceiveTeleportedAsset(T::DeliveryFee::get().into()),
+				PayFees { asset: T::DeliveryFee::get() },
```

Comment:
@yrong comments on what these fee instructions are doing please

Reply:
Also, If we are teleporting `T::DeliveryFee`, it needs to be burnt too right?

Reply:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/15854d0e2a60218ba2404ed3beb712298004fe82

> If we are teleporting T::DeliveryFee, it needs to be burnt too right?

Exactly, burnt in https://github.com/vgeddes/polkadot-sdk/blob/19f7f03635861fc31c6f5a9b2123585e44a2153a/bridges/snowbridge/pallets/system-frontend/src/lib.rs#L210